device/src/lib.rs

```diff
@@ -1,334 +1,234 @@
-                                let session = Session {
-                                    newskey: decrypt
-                                        .derive_newskey(&devnonce, &self.credentials.appkey),
-                                    appskey: decrypt
-                                        .derive_appskey(&devnonce, &self.credentials.appkey),
-                                    devaddr: DevAddr::new([
-                                        decrypt.dev_addr().as_ref()[0],
-                                        decrypt.dev_addr().as_ref()[1],
-                                        decrypt.dev_addr().as_ref()[2],
-                                        decrypt.dev_addr().as_ref()[3],
-                                    ])
-                                    .unwrap(),
-                                    fcnt: 0,
-                                };
-                                self.sm_handler = Device::joined_idle;
-                                self.sm_data = Data::Session(session);
-                            } else {
-                            }
-                        } else {
-                            panic!("Cannot possibly be decrypted already")
-                        }
-                    } else {
-                        // just some other packet, ignore
-                    }
-                    None
-                } else {
-                    self.error(radio, event)
-                }
-            }
-            _ => self.error(radio, event),
+    pub fn get_session_keys(&self) -> Option<SessionKeys> {
+        if let State::Session(session) = &self.state {
+            Some(SessionKeys::copy_from_session_data(
+                session.get_session_data(),
+            ))
+        } else {
+            None
         }
     }
 
-    fn joined_idle(&mut self, radio: &mut dyn Radio<Event = E>, event: Event) -> Option<Response> {
-        if let Data::Session(_) = self.sm_data {
-            match event {
-                Event::SendData(_) => {
-                    radio.configure_tx(
-                        14,
-                        Bandwidth::_125KHZ,
-                        SpreadingFactor::_10,
-                        CodingRate::_4_5,
-                    );
-                    let random = (self.get_random)();
-                    radio.set_frequency(self.region.get_data_frequency(random as u8));
-                    radio.send_buffer();
-                    self.sm_handler = Device::joined_sending;
-
-                    None
+    pub fn get_downlink_payload(&mut self) -> Option<(u8, Vec<u8, U256>)> {
+        let buffer = self.get_radio().get_received_packet();
+        if let Ok(parsed_packet) = lorawan_parse(buffer, C::default()) {
+            if let PhyPayload::Data(data_frame) = parsed_packet {
+                let fport = data_frame.f_port();
+                if let DataPayload::Decrypted(decrypted) = data_frame {
```

Comment:
I am surely missing something, but when will the `data_frame` be already decrypted? I don't think `lorawan_parse`/ `parse_with_factory` can produce a decrypted payload as it does not know the keys.

Reply:
Good catch! This is actually something I only noticed Friday that it was being parsed as an encrypted payload here despite the payload having been decrypted in the state machine. I'm working on a fix for this soon.

My idea is to have an enum for the send/receive buffer and to store the decrypted structure there after decryption.

Reply:
I had trouble using the same memory via enum so I just made a copy for now. I'll consider making it an enum to be an optimization for later :P